packages/xstate-solid/src/utils.ts

```diff
@@ -0,0 +1,73 @@
+import type { AnyState, AnyStateMachine, State, StateFrom } from 'xstate';
+import type { SetStoreFunction } from 'solid-js/store';
+import { produce, reconcile } from 'solid-js/store';
+import rfdc from 'rfdc';
+
+// List of keys to reconcile while merging state
+const reconcileKeys: Array<keyof AnyState> = ['context', 'value'];
+
+/**
+ * Reconcile the state of the machine with the current state of the store.
+ * Handles primitive values, arrays, and objects.
+ * Provides granular reactivity for the state of the machine in solid-js.
+ */
+export const updateState = <
+  NextState extends StateFrom<AnyStateMachine> | object,
+  UpdateState extends SetStoreFunction<NextState>
+>(
+  nextState: NextState,
+  setState: UpdateState,
+  keys?: Array<keyof NextState>
+): void => {
+  if (typeof nextState === 'object' && !!nextState) {
+    // If a list of keys to update is not provided, get keys from nextState
+    if (!keys) {
+      keys = Object.keys(nextState) as Array<keyof NextState>;
+    }
+    for (const key of keys) {
+      // Don't update functions
+      if (typeof nextState[key] === 'function') {
+        continue;
+      }
```

Comment:
In a similar fashion - this seems to be mostly useful in context of `useMachine` but probably other functions should be updated with `setProduceState`